### ISSUE TITLE
Deallocating non-allocated memory

### ISSUE THREAD

**Issue by pjshelton** [cpputest/cpputest#1498]
C:\Cardno\Oceansense\test\dateandtimetest.cpp:48: error: Failure in TEST(TDateTime_methods, DateTimeString3)
DateTimeString3:48: error:
        Deallocating non-allocated memory
   allocated at file: <unknown> line: 0 size: 0 type: unknown
   deallocated at file: <unknown> line: 0 type: delete

..
Errors (1 failures, 3 tests, 3 ran, 3 checks, 0 ignored, 0 filtered out, 14 ms)

[test.zip](https://github.com/cpputest/cpputest/files/7447723/test.zip)

Hi, I get the above error when running my test cases.  I do not know how to find out why.  I do know that it seems to have something to do with the length of the string I am comparing.  But I do not understand why.

To get the supplied code to compile, you will need to update line 9  of CMakeList.txt to point to your version of CppUTest source.  You might also need to change line 151 of DateAndTime.cpp,  depending on your includes, this may have to be an ldiv_t instead of a div_t.  

The failure is triggered by the code from line 170 on in the file DateAndTime.cpp. 
This function is taking a TDateTime and returning a std::string in a particular format. 

- Returning the date only, does not trigger a failure,
- Returning the time only does not trigger a failure
- Returning either the date and half the time, or the time and half the date does not trigger a failure;
- Only returning a string that contains both the date and the time produces the failure. 

More intriguing is the fact that when run outside the test suite, the code doesn't produce any failures.

build\test is the executable that just uses the code, and build\TDateTimeTest is the same code under the test harness. 

I would like some hints as to how I might be able to get closer to the source of the problem.

**Comment by pjshelton** [comment]
#include "generated/CppUTestGeneratedConfig.h"
#include "CppUTest/CommandLineTestRunner.h"
#include <string>
int main(int ac, char **av) {
  return CommandLineTestRunner::RunAllTests(ac, av);
}

TEST_GROUP(strcmp_test){
};
TEST(strcmp_test, string3) {
  STRCMP_EQUAL("123", std::string("123").c_str());
}
TEST(strcmp_test, string6) {
  STRCMP_EQUAL("123456", std::string("123456").c_str());
}
TEST(strcmp_test, string9) {
  STRCMP_EQUAL("123456789", std::string("123456789").c_str());
}
TEST(strcmp_test, string12) {
  STRCMP_EQUAL("123456789012", std::string("123456789012").c_str());
}
TEST(strcmp_test, string15) {
  STRCMP_EQUAL("123456789012345", std::string("123456789012345").c_str());
}
TEST(strcmp_test, string16) {
  STRCMP_EQUAL("1234567890123456", std::string("1234567890123456").c_str());
}
TEST(strcmp_test, string18) {
  const char* temp_PJ;
  std::string str_PJ = std::string("1234567890123456");
  temp_PJ = str_PJ.c_str();
  STRCMP_EQUAL("1234567890123456", temp_PJ);
}
TEST(strcmp_test, string24) {
  STRCMP_EQUAL("123456789012345678901234", "123456789012345678901234");
}


There is something I don't understand going on between your STRCMP_EQUAL and the std::string c_str().
it is find for string less then 16 char but causes problems if longer that 15 char.

**Comment by pjshelton** [comment]
Sorry.  I think I may have opened a duplicate.  #289 is all about library allocated memory and this ticket is about the library allocated memory in std::string()  It has nothing to do with STRCMP_EQUAL
because if you add MemoryLeakWarningPlugin::turnOffNewDeleteOverloads(); to the main in the above code it passes.

I have a very vague recollection that I read somewhere that std::string() has a small fixed length char[] array so that for short strings, it can just save to the internal array and does not have to make an expensive call to malloc or new.  

I am currently at a loss as to why the memory leak warning should be able to see the destruction of the std::string() but not its creation.

**Comment by basvodde** [comment]
This is probably because of a buffer that was allocated outside the test and then deleted inside the test. This will depend on the std::string implementation. The workarounds are different per specific situation. Sometimes you might want to allocate a or a few strings in main (to allocate the buffers) or sometimes create the objects on the stack in the test instead of in the TEST-GROUP. I've bumped to these before and been able to work around it most of the time but they are annoying.

**Comment by pjshelton** [comment]
I don't follow.  Sorry.  So far as I can tell I am creating the std::string() objects on the stack in the test.  I am not creating them in the TEST-GROUP.   They don't even exist until I have them wrapped them inside the STRCMP_EQUAL() macro.  

If I put them in the main, am I treating them like 'global'  variables?

**Comment by pjshelton** [comment]
okay,
in further testing,  I moved the declaration of my long std::string  to after the includes.  I then defined it as an 18 char length string in main.  I then referenced that string in the test.   So my "new" std::string outlasts the test. and does not get deleted inside the memory check.

**Comment by basvodde** [comment]
Ok, does that mean we ought to close this issue ?

**Comment by basvodde** [comment]
Yes, that is a good idea. Perhaps it could be added towards the STL section. Do you want to practice your Pull Request skills for that? :)

**Comment by pjshelton** [comment]
you got me.  erm, where do I find the githup project that is the CppUTest manual?  I have  been unable to fine the

**Comment by basvodde** [comment]
Is the borland compiler nowadays open source and available? It might be an idea to also add it to the CppUTest automated build and test

**Comment by pjshelton** [comment]
Maybe not open source but they do supply a free version (only for Win32).  it is one or two versions above the one I am using.

**Comment by basvodde** [comment]
Can it easily and automatically be installed? We have quite some builds in travis for the different compilers and environments

**Comment by basvodde** [comment]
Sorry, the CppUTest docs and site are at: https://github.com/cpputest/cpputest.github.io

**Comment by pjshelton** [comment]
to get a copy of the compiler you need to give Embarcadero your name, email, company, phone number and country. So far as I am aware it is then just a command line compiler.
However I would hold off setting it up until we are further through the changes I am supplying as if the current version that Embarcadero supply has the same problems as the version I have  you will have lots of failures while trying to compile.

**Comment by pjshelton** [comment]
basvodde commented 14 days ago
Ok, does that mean we ought to close this issue ?

It does mean that I can not test any function that produces a std::string for memory leaks. I can test it for everything else, but not leaks.

**Comment by basvodde** [comment]
I was keeping it open as a reminder that we need to add it to the docs.

How come you cannot test for memory leaks for a function that produces std::string? Isn't it so that you can only not test for leaks inside std::string ?

**Comment by pjshelton** [comment]
happy to have it left open until we both are happy this is resolved.

The reason I gave comment this morning was:
If I create an single test and in that test create a single std::string that contains a 20 character array, on exit from the test the std::string is destroyed, (I don't have an issue with that) and the memory leak infrastructure captures the release of the memory and records the destruction but has not seen the allocation.  so it complains about an imbalance. 

I can create that std::string as a global variable and initalise it in main and then use it in the test and there is no problem.

My issue is that the function I am testing creates and returns a std:string.  If I run this function outside the test, all I have is it's effects.  it returns a std::string, so I have one std::string.  I can test that this string contains what I expect is should be.  but I have not tested the code that produced the std::string for memory leaks.  because that code ran outside of the test.  when I try to run the function inside the test the std::string is created in the test and then we have out problem

**Comment by basvodde** [comment]
But any other memory leaks would be checked, right? It is only that you don't test the std::string itself?

**Comment by pjshelton** [comment]
I am really sorry. I don't seem to  be expressing my self at all well.
my function signature is

> std::string func();

So when this function is called, a std::string is created in the calling scope.  
my testing shows that on my machine, using gcc (MinGW-W64 x86_64-ucrt-posix-seh, built by Brecht Sanders) 11.2.0 (see https://winlibs.com), that creating and destroying a std::string inside the scope of a memory test results in the MemoryLeakDetection catching a Deallocating non-allocated memory error. 

> #include "generated/CppUTestGeneratedConfig.h"
> #include "CppUTest/CommandLineTestRunner.h"
> #include <string>
> int main(int ac, char **av) {
> return CommandLineTestRunner::RunAllTests(ac, av);
> }
> TEST_GROUP(strcmp_test){
> };
> TEST(strcmp_test, string16) {
> std::string("1234567890123456");
> }

when I compile and run the above code I get the error.  All I am doing is creating an un-named std:string.

When I use my function it creates a std::string, and when I call in in the scope of a test I get the above error and when I call it outside the test and just test that the returned string contains what I want it to, I am not testing whether the function leaks because the running of the function happens outside of the test.  

So I am not  quite sure  what  I am to write in the docs as it seems a bit harsh to call out a particular compiler/library and say that for a particular use case one can not do a proper test for memory leaks.  But that is what I am seeing.

I hope that the above is not total gibberish.
